datafusion/core/tests/fuzz_cases/aggregation_fuzzer/query_builder.rs

```diff
@@ -0,0 +1,371 @@
+    ///
+    /// ```text
+    ///   SELECT aggr FROM t;
+    /// ```
+    ///
+    no_grouping: bool,
+
+    // ====================================
+    // Aggregation function settings
+    // ====================================
+    /// Aggregate functions to be used in the query
+    /// (function_name, is_distinct)
+    aggregate_functions: Vec<(String, bool)>,
+
+    /// Possible columns for arguments in the aggregate functions
+    ///
+    /// Assumes each
+    arguments: Vec<String>,
+}
+
+impl QueryBuilder {
+    pub fn new() -> Self {
+        Self {
+            no_grouping: true,
+            max_group_by_columns: 3,
+            ..Default::default()
+        }
+    }
+
+    /// return the table name if any
+    pub fn table_name(&self) -> &str {
+        &self.table_name
+    }
+
+    /// Set the table name for the query builder
+    pub fn with_table_name(mut self, table_name: impl Into<String>) -> Self {
+        self.table_name = table_name.into();
+        self
+    }
+
+    /// Add a new possible aggregate function to the query builder
+    pub fn with_aggregate_function(
+        mut self,
+        aggregate_function: impl Into<String>,
+    ) -> Self {
+        self.aggregate_functions
+            .push((aggregate_function.into(), false));
+        self
+    }
+
+    /// Add a new possible `DISTINCT` aggregate function to the query
+    ///
+    /// This is different than `with_aggregate_function` because only certain
+    /// aggregates support `DISTINCT`
+    pub fn with_distinct_aggregate_function(
+        mut self,
+        aggregate_function: impl Into<String>,
+    ) -> Self {
+        self.aggregate_functions
+            .push((aggregate_function.into(), true));
+        self
+    }
+
+    /// Set the columns to be used in the group bys clauses
+    pub fn set_group_by_columns<'a>(
+        mut self,
+        group_by: impl IntoIterator<Item = &'a str>,
+    ) -> Self {
+        self.group_by_columns = group_by.into_iter().map(String::from).collect();
+        self
+    }
+
+    /// Add one or more columns to be used as an argument in the aggregate functions
+    pub fn with_aggregate_arguments<'a>(
+        mut self,
+        arguments: impl IntoIterator<Item = &'a str>,
+    ) -> Self {
+        let arguments = arguments.into_iter().map(String::from);
+        self.arguments.extend(arguments);
+        self
+    }
+
+    /// Add max columns num in group by(default: 3), for example if it is set to 1,
+    /// the generated sql will group by at most 1 column
+    #[allow(dead_code)]
+    pub fn with_max_group_by_columns(mut self, group_by_columns: usize) -> Self {
+        self.max_group_by_columns = group_by_columns;
+        self
+    }
+
+    /// Add sort keys of dataset if any, then the builder will generate queries basing on it
+    /// to cover the sort-optimization cases
+    pub fn with_dataset_sort_keys(mut self, dataset_sort_keys: Vec<Vec<String>>) -> Self {
```

Comment:
Another issue I encounter is that it is not easy to control the ordering of the columns by this API.

For example, we can generate query like this, but there is no easy way to determine the ordering of the group by columns.

```rust
pub fn generate_multi_group_query(&self) -> String {
        let group_by_cols = self.random_group_by(Some(2));
        format!(
            "SELECT sum({}) FROM {} GROUP BY {}, {}",
            group_by_cols[0], self.table_name, group_by_cols[0], group_by_cols[1]
        )
    }
```

Reply:
The workaround I did is to limit the number column I provided like this


```rust
let columns = get_supported_types_columns(rng.gen());

    for column in columns {
        let sort_key = column.name.to_string();
        let data_gen_config = DatasetGeneratorConfig {
            columns: vec![column],
            rows_num_range,
            sort_keys_set: vec![vec![sort_key]],
        };
```

Reply:
If we can assign ordering by index and differentiate group by columns vs normal columns, probably is better? 

```rust
let data_gen_config = DatasetGeneratorConfig {
            columns,
            group_by_columns,
            rows_num_range,
            sort_keys_set,
            group_sort_keys_set: vec![vec![0, 1]], // first 2 group by columns are ordered
        };
```

Reply:
If I don't misunderstand, the actual requirement is like that? 
- We need to test sorted group by optimization path
- But now, it is hard to randomly generate the right sort keys, due to the we need to restrict the sort keys to be the group by keys?

Reply:
For example, the dataset has columns `[a,b,c,d,e]`.

And we randomly pick `[a,b]` as the `sort key`.

But unfortunately, only `[c,d,e]` picked as the `group by key`.

So finally, the testcases can not cover the sort-optimization?

Reply:
The sql picking logic also has some problems now as I noticed.

For example, assume dataset `t1` sorted by `[a, b]`, and `t2` sorted by `[c,d]`.

It is possible that
- we randomly picked sql `select xxx from xxx group by c,d` for `t1`
- but `select xxx from xxx group by a,b` for `t2`

I am also fixing this.

Reply:
@jayzhan211 I am not sure if I understand the cases rightly?

Actually it will specially generate queries grouping by the columns in the `sort_keys_set`.

For example,
If we define a dataset configs like:
```text
columns: [a, b, c, d, e]
sort_keys: [a,b]
```

And the query builder will firstly generate sql with `all columns`, and in this round the generated sqls are possibly not grouping on the sort keys:
```
SELECT aggr from t GROUP BY c
SELECT aggr from t GROUP BY c,d
SELECT aggr from t GROUP BY e
...
```

But the special round  generating sql gouping on the sort keys([a,b] here) exist:
```
SELECT aggr from t GROUP BY a
SELECT aggr from t GROUP BY b
SELECT aggr from t GROUP BY a,b
```

Reply:
However, the sql picking logic indeed has some problems now as I noticed.

For example, assume dataset t1 sorted by [a, b], and t2 sorted by [c,d].

It is possible that
- we randomly picked sql select xxx from xxx group by c,d for t1
- but select xxx from xxx group by a,b for t2